Integrate Validation and Success Feedback for Link Investigation

Heuy-kt opened the pull request:
This PR enhances the Detective Z application with improved user input handling, validation feedback, and success messaging. It integrates backend validation with the frontend to provide a seamless user experience for link investigations.

---

#### Key Changes:
1. **Enhanced `index.html`**:
   - Added a Thymeleaf-enabled `<form>` for submitting links to the `/investigate` endpoint.
   - Included an error handling section to dynamically display validation errors (`${errors}`).
   - Introduced a `Reset` button for clearing form inputs.

2. **Created `investigate.html`**:
   - A minimal success page with an `<h1>` message confirming successful link validation.
   - Set the page title to "Link is good to go."

3. **Updated `InvestigationController`**:
   - Enhanced `/investigate` POST endpoint with `BindingResult` for handling validation errors.
   - Added `Model` attributes to pass errors and success messages to views.
   - Ensured the POST endpoint redirects to the `index` page with error feedback if validation fails.
   - Redirects to the `investigate` page upon successful validation.

---

#### Testing:
- Tested form submission with valid and invalid inputs to ensure proper error and success handling.
- Verified the `index.html` and `investigate.html` pages display the expected content based on validation results.
- Checked that the `Reset` button clears the form fields.

---

This PR introduces a crucial validation mechanism that aligns with the application's goal of promoting safe and legitimate links. 🎉

aalamu commented on src/main/java/com/fleencorp/detectivez/controller/InvestigationController.java:
```diff
@@ -17,8 +20,14 @@ public String investigate() {
-  public String investigate(@Valid @RequestBody InvestigateDto investigateDto) {
+  @PostMapping(value = "investigate")
+  public String investigate(@Valid @RequestBody InvestigateDto investigateDto, BindingResult bindingResult, Model model) {
+    if (bindingResult.hasErrors()){
```
We should learn to apply a space between the closing bracket of a condition and the brance that follows it.

Heuy-kt replied:
Noted. **lol** i just never liked that from scratch.

aalamu commented on src/main/java/com/fleencorp/detectivez/controller/InvestigationController.java:
```diff
@@ -17,8 +20,14 @@ public String investigate() {
+      model.addAttribute("errors", bindingResult.getAllErrors());
+      return "index";
+    }
+    model.addAttribute("message", "success");
```
The message should be as written in the ticket

https://github.com/aalamu/detective-z/issues/7

aalamu commented on src/main/resources/templates/index.html:
```diff
@@ -1,10 +1,36 @@
+<h1>Investigate!</h1>
+<form action="/investigate" method="post">
+    <!-- Text input for link -->
+    <label>input link:</label><br>
```
1. Form labels should be title case.
2. A form label should have an ID and a matching ID in the form control or field.

aalamu commented on src/main/resources/templates/index.html:
```diff
@@ -1,10 +1,36 @@
+<p>I am Detective Z and I will inform you about what is not legit.</p>
+
+<h1>Investigate!</h1>
+<form action="/investigate" method="post">
```
```enctype``` attribute should be set in the form element.

aalamu commented on src/main/resources/templates/index.html:
```diff
@@ -1,10 +1,36 @@
+    <title>Detective Z</title>
+</head>
+<body>
+<h1>Your Security Companion</h1>
```
```h1``` and ```p``` elements should be indented to the right from the ```body``` element to make templates clean and more readable.

Line space should be applied where necessary.

```form``` elements and its children should be indented also.

aalamu commented on src/main/resources/templates/index.html:
```diff
@@ -1,10 +1,36 @@
+<!--error section-->
+
+    <div th:if="${errors != null}">
+        <h2>Errors!</h2>
```
```Errors!``` can be replaced with ```Errors```

aalamu commented on src/main/resources/templates/index.html:
```diff
@@ -1,10 +1,36 @@
+<form action="/investigate" method="post">
+    <!-- Text input for link -->
+    <label>input link:</label><br>
+    <input type="text" id="link" name="link" placeholder="Enter link for investigation" required><br><br>
```
```input``` element should be self closed.

aalamu replied:
```br``` is a self-closing tag and should be applied.

two ```br``` elements should be moved to the next line.

aalamu commented on src/main/resources/templates/investigate.html:
```diff
@@ -0,0 +1,9 @@
+<!doctype html>
+<html>
+<head>
+    <title>Link is good to go</title>
```
```title``` element text content can be better for example.

```Investigation Result```

aalamu commented on src/main/resources/templates/investigate.html:
```diff
@@ -0,0 +1,9 @@
+    <title>Link is good to go</title>
+</head>
+<body>
+    <h1>Success</h1>
```
**Success** as a visible heading is vague and should be improved with the purpose of the page.

#### For example:
1. Investigation Result
2. Analysis Result
3. etc.

A paragraph should be added to tell user more information about investigation and analysis.

aalamu reviewed (changes requested):
We need to apply the changes and make improvements

aalamu commented:
Hi @Heuy-kt ,

There's a merge conflict that needs to be resolved in addition to needed changes

Heuy-kt commented:
**Title:** Refactor HTML and Controller Logic for Improved Validation and User Experience  

**Description:**  
This PR includes multiple improvements and refinements across HTML templates and controller logic to enhance code readability, validation handling, and user experience.  

**Changes Made:**  
1. Removed the exclamation mark from the investigate header for better tone consistency.  
2. Capitalized the "post" method to "POST" to align with HTTP standards.  
3. Added the `enctype` attribute to the form and set it to "application/x-www-form-urlencoded".  
4. Changed the `link` label to title case for consistency.  
5. Added a `for` attribute to the `link` label and linked it to the ID of the input field for accessibility.  
6. Removed the exclamation mark in the `<h2>` Errors header for clarity.  
7. Updated the `title` in `investigate.html` from "Link is good to go" to "Investigation Result" for accuracy.  
8. Replaced the "Success" header with "Link Registration Result" to provide a more descriptive outcome.  
9. Added additional headers and a paragraph to improve content structure.  
10. Added proper spacing between conditional statement brackets and their block implementation for better readability.  
11. Updated the value of `message` in the model attributes from "success" to "Link investigated successfully" as specified in the ticket.  

**Why This Change is Needed:**  
These changes were made to address inconsistencies in formatting, improve accessibility, ensure better compliance with coding standards, and provide users with a clearer and more professional interface and feedback during link investigations.

aalamu commented:
Hi @Heuy-kt ,

Did you tested the application, forms and validations and it works?

Heuy-kt commented:
The code works. and on post man the expected pages are displayed. but theres been an issue with the `investigate.link` message. 
![image](https://github.com/user-attachments/assets/b20b037b-a12a-4834-ba4f-5f2527bf5697)
